### ISSUE TITLE
Not matching correctly the grammar

### ISSUE THREAD

**Issue by andrfgs** [antlr/intellij-plugin-v4#436]
I'm using ANTLR4 with the Intellij plugin but seems like the plugin is broken at least for Intellij 2020.3/Clion 2020.3 

I'm using Python3 grammar from https://github.com/antlr/grammars-v4 

I generated 2 parser programs: a Java version and a C++ version, using the appropriate grammars adapted for each target. Both programs are working correctly and parsing as they should (though the C++ target `SEGFAULTS` when using an empty input file of only 2 `newlines`, but anyways). But the plugin on `ANTLR Preview` doesn't match the exact same file with the exact same starting rule (`file_input`)

Basically this is the input I am testing:
```
import a
```

This is accepted by both the Java and C++ program using the `file_input` rule but the plugin fails with:
`line 1:6 extraneous input ' ' expecting NAME`

For some reason, the plugin is parsing the white space as a token, but the parser programs are doing just fine. What is going on here?

**Comment by bjansen** [comment]
As explained [in the readme](https://github.com/antlr/intellij-plugin-v4#limitations):

>The ANTLR Preview window is based on a grammar interpreter, not on the actual generated parser. This means that things like actions and predicates will not be evaluated during live preview, because the interpreter is language agnostic.
For the same reasons, if your parser and/or lexer classes extend a custom implementation of the base parser/lexer classes, your custom code will not be run during live preview.

The Python3 grammar has overridden `nextToken()` to match Python indents correctly. This custom code can unfortunately not be run from the Preview window, that's why you get lexing errors on whitespace.

**Comment by andrfgs** [comment]
I see. I'm terribly sorry then. I was assuming this could be a bug on newer versions of IntelliJ/CLion since I found some issues here about it, but wasn't sure of them.

By the way, this grammar works on `ANTLRWorks2`, so I'm assuming the program generates the Java Parser unlike the plugin? The utility I was seeking was the syntax tree, but I'll use `ANTLRWorks2` or `grun` instead. Is there a specific reason the plugin wouldn't generate the grammar by itself and instead uses that builtin grammar interpreter?

I'll look later into the IntelliJ plugin source and check If I can manage to apply this modification in a fork.

**Comment by bjansen** [comment]
I'm not sure why the plugin was designed this way, I think @parrt can answer that.

I also think ANTLRWorks generates and runs the actual parser, that's why it can also *debug* grammars. But I you want to support any target, not just Java, display the list of tokens, matched rules, ambiguities, profiling results etc. then I think you have no other choice than to interpret the grammar instead of running the actual parser. I'm no ANTLR expert though, so I might be wrong :)

**Comment by parrt** [comment]
Yep, @bjansen has it right.  Old ANTLRWorks executed parsers that sent events to a listener.  The plugin is now interpreting, not executing generated code so can't handle support code like tweaks to nextToken().

**Comment by bjansen** [comment]
@parrt is there a specific reason for using an interpreter rather than a generated parser? A lot of people seem to be annoyed by this.

**Comment by andrfgs** [comment]
By the way, I want to state that I am very grateful for ANTLR as it is and I really thank all the contributors here that devote their time improving this project. I'm not demanding something by any means, I was simply thinking this was a bug in the plugin.

> I also think ANTLRWorks generates and runs the actual parser, that's why it can also debug grammars. But I you want to support any target, not just Java, display the list of tokens, matched rules, ambiguities, profiling results etc. then I think you have no other choice than to interpret the grammar instead of running the actual parser. 

Unfortunately, I have zero experience in compilers and LL grammars, so changing much of the Python grammar to be target independent is beyond my knowledge. So my solution is to use the 2 provided Python grammars (one for C++ and one for Java) and change them to my specifications. I'm working on a transpiler to be exact.

**Comment by parrt** [comment]
Yes, no way for plugin to know how to build, launch, test code in all of these 10 languages. Could be done but plugin is for testing grammars not your actions.

**Comment by bjansen** [comment]
> plugin is for testing grammars not your actions.

Right, but sometimes code *needs* to be executed to test the grammar correctly, like the indentation thing in the Python3 grammar.

**Comment by andrfgs** [comment]
What would I need to change in the Python3 grammar to support the indentation? I still don't understand the problem, is it due to the grammar in itself?

**Comment by parrt** [comment]
@bjansen yep, but I am unable to easily support actions. Imagine the actions required data files or something. Imaging a java package specifier. That alone could screw up my execution until I store in proper directory structure etc..  each language has these quirks.  It's not solvable in general and I accepted the limitation.

**Comment by bjansen** [comment]
> What would I need to change in the Python3 grammar to support the indentation? I still don't understand the problem, is it due to the grammar in itself?

Yes, the Preview window will *not* execute actions/predicates in `Python3.g4` because it doesn't run Java code, it just interprets the grammar. In your case, the predicate `{atStartOfInput()}?` will be ignored by the interpreter:

```antlr
NEWLINE
 : ( {atStartOfInput()}?   SPACES
   | ( '\r'? '\n' | '\r' | '\f' ) SPACES?
```
As a consequence, spaces will always match `NEWLINE`. 
If you run the real generated parser, `{atStartOfInput()}?` will evaluate to `false` because your space is not at the start of input. The `NEWLINE` rule will thus not match, but `SKIP_` probably will:

```antlr
SKIP_
 : ( SPACES | COMMENT | LINE_JOINING ) -> skip
 ;
```

So there's no easy way to make a Python grammar work correctly in the Preview window I guess.
